purenix/src/Nix/Identifiers.hs

```diff
@@ -0,0 +1,102 @@
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+-- | This module defines the conversions from purescript identifiers to (pure)nix identifiers.
+-- Nix's rules for identifiers are strictly more lenient than purescripts (see specifications below), since nix allows - in non-head characters.
+-- This makes it so that in general, we can just convert identifiers.
+-- However, there are a number of edge cases that need to be handled:
+--  - generated dictionary names will contain '$'
+--  - module names contain '.'
+--  - we should not shadow "builtins", or identifiers starting with "__". There are of course more identifiers, that could techincally be shadowed, but they tend to overlap with purescript's.
+--
+-- In general, once a purescript construct has been converted into a Key/Var, no more work is required.
+-- That is, after turning them into Text, they can be used directly by the printer, no more work needs to be done to escape them.
+--
+-- Purescript identifiers:
+-- https://github.com/purescript/purescript/blob/master/lib/purescript-cst/src/Language/PureScript/CST/Lexer.hs#L689
+-- Nix identifiers:
+-- https://github.com/cstrahan/tree-sitter-nix/blob/master/src/grammar.json#L19
+module Nix.Identifiers
+  ( Var (..),
+    mkVar,
+    numberedVars,
+    Key (..),
+    identKey,
+    stringKey,
+    moduleKey,
+    binderKey,
+    numberedKeys,
+  )
+where
+
+import Data.Maybe
+import Data.Set (Set)
+import qualified Data.Set as Set
+import Data.String
+import Data.Text (Text)
+import qualified Data.Text as T
+import qualified Language.PureScript as PS
+import qualified Language.PureScript.PSString as PS
+
+-- TODO rename to Binder, since this can occur in the LHS of a let-binding
+newtype Var = UnsafeVar {unVar :: Text}
+  deriving newtype (IsString, Eq, Show)
+
+identToText :: PS.Ident -> Text
+identToText (PS.Ident t) = t
+-- GenIdent is only used in PureScript for "unnamed" instances.
+-- Originally, in PureScript, all instances needed to be named:
+-- https://github.com/purescript/documentation/blob/master/language/Differences-from-Haskell.md#named-instances
+-- This was relaxed in 0.14.2:
+-- https://github.com/purescript/purescript/pull/4096
+identToText (PS.GenIdent mvar n) = fromMaybe "__instance" mvar <> T.pack (show n)
+identToText PS.UnusedIdent = error "impossible"
+
+mkVar :: PS.Ident -> Var
+mkVar = UnsafeVar . removeDollarSigns . tickKeywords . identToText
+  where
+    tickKeywords w
+      | Set.member w keywords || T.isPrefixOf "__" w = w <> "-"
+      | otherwise = w
+    removeDollarSigns w =
+      T.map (\c -> if c == '$' then '-' else c) $
+        if T.isPrefixOf "$" w then "S" <> w else w
```

Comment:
What sort of PureScript code did you convert to get it to give you variable names with a leading `$`?  I tried playing around with a few things, but I couldn't figure out anything.

I was wondering if we'd have some sort of problem where PureScript gave us an identifier like `$hello`, so we convert it to an identifier like `Shello`, but the user also defines a data constructor in the same file called `Shello` (like `data Foo = Shello`).

Reply:
I could've sworn they showed up in the names of dictionaries, but grepping the code base, all of them seem to occur as `$__unused`, for example
```
Data.Bounded/default.nix:      "Ord0" = $__unused: module."Data.Ord".ordProxy3;
Data.Bounded/default.nix:      "Ord0" = $__unused: module."Data.Ord".ordProxy2;
Data.Bounded/default.nix:      "Ord0" = $__unused: module."Data.Ord".ordProxy;
Data.Bounded/default.nix:      "Ord0" = $__unused: module."Data.Ord".ordOrdering;
```
Since we always escape things with two leading underscores anyway, I think it's safe to just drop the $ like I've done in f421e5d. We'll change it back if it causes issues.